packages/next/src/lib/metadata/generate/icons.tsx

```diff
@@ -0,0 +1,76 @@
+import type { ResolvedMetadata } from '../types/metadata-interface'
+import type { Icon, IconDescriptor } from '../types/metadata-types'
+
+import React from 'react'
+import { resolveAsArrayOrUndefined } from './utils'
+
+const resolveUrl = (url: string | URL) =>
+  typeof url === 'string' ? url : url.toString()
+
+function IconDescriptorLink({ icon }: { icon: IconDescriptor }) {
+  const { url, rel = 'icon', ...props } = icon
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (!url) {
+      throw new Error('`url` in required for descriptor in `metadata.icons`')
+    }
+  }
```

Comment:
Can we move all the validation code to one unified place after resolving the metadata? Instead of checking them separately. I think that will make things easier to organize and read.

Reply:
I'll remove this for now, and we can bring more checking back when the typing is improved.
since for the resolved typing, those icon fields have various forms `"Icon | Icon[] | undefined"`, would be nice if it's always array or undefined